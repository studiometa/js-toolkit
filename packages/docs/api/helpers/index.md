---
sidebar: true
---

# Helpers

The following helper functions can be used to achieve advanced setup of your application.

## App helper

The following helpers can help you bootstrap the root class of your application:

- [createApp](./createApp.html)

## Lazy import helpers

Some components might not need to be imported and instantiated immediately on page load. The following functions will help you define custom scenarios for when to import these components.

- [importOnInteraction](./importOnInteraction.html)
- [importWhenIdle](./importWhenIdle.html)
- [importWhenVisible](./importWhenVisible.html)
