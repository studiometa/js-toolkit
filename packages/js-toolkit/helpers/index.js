export { default as createApp } from './createApp.js';
export { default as getInstanceFromElement } from './getInstanceFromElement.js';
export { default as importOnInteraction } from './importOnInteraction.js';
export { default as importWhenIdle } from './importWhenIdle.js';
export { default as importWhenVisible } from './importWhenVisible.js';
