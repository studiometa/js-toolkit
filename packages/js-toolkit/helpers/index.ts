export * from './getDirectChildren.js';
export { default as createApp } from './createApp.js';
export { default as getClosestParent } from './getClosestParent.js';
export { default as getInstanceFromElement } from './getInstanceFromElement.js';
export { default as importOnInteraction } from './importOnInteraction.js';
export { default as importWhenIdle } from './importWhenIdle.js';
export { default as importWhenVisible } from './importWhenVisible.js';
export { default as importOnMedia } from './importOnMedia.js';
