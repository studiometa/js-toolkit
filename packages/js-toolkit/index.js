export { default as Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';
