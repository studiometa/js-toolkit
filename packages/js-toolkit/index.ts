export { Base } from './Base/index.js';
export * from './decorators/index.js';
export * from './helpers/index.js';
export * from './services/index.js';

export type {
  BaseTypeParameter,
  BaseConstructor,
  BaseAsyncConstructor,
  BaseConfig,
} from './Base/index.js';

export type { BaseInterface, BaseDecorator } from './Base/types.js';
