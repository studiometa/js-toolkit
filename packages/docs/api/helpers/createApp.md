---
outline: deep
---

# createApp

Use the `createApp` function to instantiate your application on page load while being able to use the instance in other files.

## Usage

```js {1,4,6} twoslash
import { createApp } from '@studiometa/js-toolkit';
import App from './App.js';

const useApp = createApp(App);

useApp().then((app) => console.log(app));
```

**Parameters**

- `App` (`typeof Base`): the root class for your app

The second parameter can either be one of the following:

1. `rootElement` (`HTMLElement`): the root element for your app, defaults to `document.body`
2. `options` (`Object`): an object to configure more advanced options
   - `options.root` (`HTMLElement`): the root element for your app, defaults to `document.body`
   - `options.breakpoints` (`Record<string, string>`): a list of breakpoints to confgure the [`useResize` service](/api/services/useResize)
   - `options.blocking` (`boolean`): wether to enable the queue mechanism for the internals of the framework or not, defaults to `false`
   - `options.prefix` (`string`): customize the prefix used to search for components with custom element tag names, defaults to `tk`
   - `options.attributes` (`{ component: string, option: string, ref: string }`): the HTML attributes to use for the [HTML API](/api/html/), defaults to `data-component`, `data-option` and `data-ref`

**Return value**

- `useApp` (`() => Promise<Base>`): a function returning a promise which resolves to the app instance

## Examples

### Basic usage

Use the `createApp` function to export your app:

```js {1,9} twoslash
import { Base, createApp } from '@studiometa/js-toolkit';

class App extends Base {
  static config = {
    name: 'App',
  };
}

export default createApp(App);
```

And import it in other files to access your app instance when it is mounted:

```js {1,4-5}
import useApp from './app.js';

(async () => {
  const app = await useApp();
  console.log(app.$el); // document.body
})();
```

### Configure breakpoints

You can configure the breakpoints used by your application by passing an options object as second parameter:

```js twoslash
import { createApp } from '@studiometa/js-toolkit';
import App from './App.js';

export default createApp(App, {
  breakpoints: {
    xxs: '0px',
    xs: '480px',
    s: '768px',
    m: '1024px',
    l: '1280px',
    xl: '1440px',
    xxl: '1920px',
    xxxl: '2560px',
  },
});
```

### Configure the queue mechanism of the app

By default, the app will be created with a queue mechanism based on the [`SmartQueue` class](/utils/SmartQueue.html). This queue helps reduce JavaScript's blocking time when instantiating the application and all of its children by releasing the current loop every 40 ms.

If you wish to use a blocking strategy, you can disable the queue by passing the `blocking` option to `true`:

```js twoslash
import { createApp } from '@studiometa/js-toolkit';

export default createApp(App, {
  blocking: true,
});
```

### Configure the HTML API attributes

The default attributes used to interact with the DOM are `data-component`, `data-option-...` and `data-ref`. These can be configured by specifying the `attributes` property of the options parameter:

```js twoslash
import { createApp } from '@studiometa/js-toolkit';
import App from './App.js';

export default createApp(App, {
  attributes: {
    component: 'tk-is',
    option: 'tk-opt',
    ref: 'tk-ref',
  },
});
```

::: warning W3C validation
Be aware that using other attributes than `data-...` attributes will probably make your HTML fail W3C validation tests.
:::
