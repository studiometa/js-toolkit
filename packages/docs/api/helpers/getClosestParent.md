# getClosestParent

Use the `getClosestParent` function to get the closest instance of a given parent component.

## Usage

```js {1,9,15}
import { Base, getClosestParent } from '@studiometa/js-toolkit';
import Child from './Child.js';

class Parent extends Base {
  static config = {
    name: 'Parent',
    components: {
      Child,
      Parent, // Useful for recursive components only
    },
  };

  onChildClick(index, event) {
    const childInstance = this.$children.Child[index];
    const parent = getClosestParent(childInstance, Parent);

    if (parent === this) {
      event.preventDefault();
    }
  }
}
```

**Parameters**

- `childInstance` (`Base`): the instance from which to look up for the parent
- `parentConstructor` (`typeof Base`): the constructor of the parent component

**Return value**

- `Base | null`: the instance of the closes parent, `null` if no parent matching the constructor was found
