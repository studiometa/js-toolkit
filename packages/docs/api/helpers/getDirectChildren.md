---
outline: deep
---

# getDirectChildren

Use the `getDirectChildren` function to get a list components which are direct descendants of the given parent instance. This function is helpful when working with nested components which declare themselves as children.

:::tip
If you need to only check if an instance is a direct descendant of another instance, prefer the [`isDirectChild` helper function](/api/helpers/isDirectChild.md) which will return a `boolean` directly.
:::

## Usage

```js {1,9,16} twoslash
import { Base, getDirectChildren } from '@studiometa/js-toolkit';
import Child from './Child.js';

class Parent extends Base {
  static config = {
    name: 'Parent',
    components: {
      Child,
      Parent, // Useful for recursive components only
    },
  };

  onChildClick({ target, event }) {
    const directChildren = getDirectChildren(this, 'Parent', 'Child');

    if (directChildren.includes(target)) {
      event.preventDefault();
    }
  }
}
```

**Parameters**

- `parentInstance` (`Base`): the target element
- `parentName` (`string`): the name of the recursive parent as specified in the `config.components` object.
- `childrenName` (`string`): the name of the children components as specified in the `config.components` object.

**Return value**

- `Base[]`: a list of the direct child components corresponding to the given `childrenName`
