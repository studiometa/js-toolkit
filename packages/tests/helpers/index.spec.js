import * as helpers from '@studiometa/js-toolkit/helpers';

test('helpers exports', () => {
  expect(Object.keys(helpers)).toMatchSnapshot();
});
