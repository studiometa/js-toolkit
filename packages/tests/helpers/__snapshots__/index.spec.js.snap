// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers exports 1`] = `
Array [
  "createApp",
  "getDirectChildren",
  "getInstanceFromElement",
  "importOnInteraction",
  "importWhenIdle",
  "importWhenVisible",
  "isDirectChild",
]
`;
